From 9d3815aec27691b1e644251ca2c87b2059e404a5 Mon Sep 17 00:00:00 2001
From: dianlujitao <dianlujitao@lineageos.org>
Date: Thu, 1 Feb 2018 22:51:56 +0800
Subject: [PATCH 089/178] input: fingerprint: goodix: Honor proximity state

Implement a poor man's pocket mode.

 * Receive proximity state from userspace
 * Disable IRQ while proximity state is near

Based on fluxi's original idea (https://review.lineageos.org/#/c/152043/),
adapted with goodix driver.

Change-Id: Icb0561ea16bce28a64715b2eac3c9edd1fb90354
---
 drivers/input/fingerprint/goodix/gf_spi.c | 52 +++++++++++++++++++++++
 drivers/input/fingerprint/goodix/gf_spi.h |  1 +
 2 files changed, 53 insertions(+)

diff --git a/drivers/input/fingerprint/goodix/gf_spi.c b/drivers/input/fingerprint/goodix/gf_spi.c
index 4efc3e0615ac..baf31edfb878 100644
--- a/drivers/input/fingerprint/goodix/gf_spi.c
+++ b/drivers/input/fingerprint/goodix/gf_spi.c
@@ -660,6 +660,10 @@ static int goodix_fb_state_chg_callback(struct notifier_block *nb,
 			if (gf_dev->device_available == 1) {
 				gf_dev->fb_black = 1;
 				gf_dev->wait_finger_down = true;
+				/* Disable IRQ when screen turns off,
+				 * only if proximity sensor is covered */
+				if (gf_dev->proximity_state)
+					gf_disable_irq(gf_dev);
 #if defined(GF_NETLINK_ENABLE)
 				msg = GF_NET_EVENT_FB_BLACK;
 				sendnlmsg(&msg);
@@ -670,8 +674,11 @@ static int goodix_fb_state_chg_callback(struct notifier_block *nb,
 			}
 			break;
 		case FB_BLANK_UNBLANK:
+		case FB_BLANK_NORMAL:
 			if (gf_dev->device_available == 1) {
 				gf_dev->fb_black = 0;
+				/* Unconditionally enable IRQ when screen turns on */
+				gf_enable_irq(gf_dev);
 #if defined(GF_NETLINK_ENABLE)
 				msg = GF_NET_EVENT_FB_UNBLACK;
 				sendnlmsg(&msg);
@@ -693,6 +700,41 @@ static struct notifier_block goodix_noti_block = {
 	.notifier_call = goodix_fb_state_chg_callback,
 };
 
+static ssize_t proximity_state_set(struct device *dev,
+	struct device_attribute *attr, const char *buf, size_t count)
+{
+	struct gf_dev *gf_dev = dev_get_drvdata(dev);
+	int rc, val;
+
+	rc = kstrtoint(buf, 10, &val);
+	if (rc)
+		return -EINVAL;
+
+	gf_dev->proximity_state = !!val;
+
+	if (gf_dev->fb_black) {
+		if (gf_dev->proximity_state) {
+			/* Disable IRQ when screen is off and proximity sensor is covered */
+			gf_disable_irq(gf_dev);
+		} else {
+			/* Enable IRQ when screen is off and proximity sensor is uncovered */
+			gf_enable_irq(gf_dev);
+		}
+	}
+
+	return count;
+}
+static DEVICE_ATTR(proximity_state, S_IWUSR, NULL, proximity_state_set);
+
+static struct attribute *attrs[] = {
+	&dev_attr_proximity_state.attr,
+	NULL
+};
+
+static const struct attribute_group attr_group = {
+	.attrs = attrs,
+};
+
 static struct class *gf_class;
 #if defined(USE_SPI_BUS)
 static int gf_probe(struct spi_device *spi)
@@ -801,12 +843,22 @@ static int gf_probe(struct platform_device *pdev)
 	gf_dev->notifier = goodix_noti_block;
 	fb_register_client(&gf_dev->notifier);
 
+	dev_set_drvdata(&gf_dev->spi->dev, gf_dev);
+
+	status = sysfs_create_group(&gf_dev->spi->dev.kobj, &attr_group);
+	if (status) {
+		pr_err("%s: Failed to create sysfs\n", __func__);
+		goto error_sysfs;
+	}
+
 	wake_lock_init(&fp_wakelock, WAKE_LOCK_SUSPEND, "fp_wakelock");
 
 	pr_info("version V%d.%d.%02d\n", VER_MAJOR, VER_MINOR, PATCH_LEVEL);
 
 	return status;
 
+error_sysfs:
+	sysfs_remove_group(&gf_dev->spi->dev.kobj, &attr_group);
 #ifdef AP_CONTROL_CLK
 gfspi_probe_clk_enable_failed:
 	gfspi_ioctl_clk_uninit(gf_dev);
diff --git a/drivers/input/fingerprint/goodix/gf_spi.h b/drivers/input/fingerprint/goodix/gf_spi.h
index 1c45bf1882df..8cc5a684ebb0 100644
--- a/drivers/input/fingerprint/goodix/gf_spi.h
+++ b/drivers/input/fingerprint/goodix/gf_spi.h
@@ -139,6 +139,7 @@ struct gf_dev {
 	char fb_black;
 	char wait_finger_down;
 	struct work_struct work;
+	bool proximity_state; /* 0:far 1:near */
 };
 
 int gf_parse_dts(struct gf_dev* gf_dev);
-- 
2.49.0

